crates/datasources/src/mongodb/exec.rs

```diff
@@ -14,35 +14,25 @@ use datafusion::physical_plan::{
 };
 use datafusion_ext::metrics::DataSourceMetricsStreamAdapter;
 use futures::{Stream, StreamExt};
-use mongodb::bson::{Document, RawDocumentBuf};
-use mongodb::{options::FindOptions, Collection};
+use mongodb::bson::RawDocumentBuf;
+use mongodb::Cursor;
 use std::any::Any;
 use std::fmt;
 use std::pin::Pin;
 use std::sync::Arc;
 use std::task::{Context, Poll};
 
-/// Field name in mongo for uniquely identifying a record. Some special handling
-/// needs to be done with the field when projecting.
-const ID_FIELD_NAME: &str = "_id";
-
 #[derive(Debug, Clone)]
 pub struct MongoBsonExec {
-    schema: Arc<ArrowSchema>,
-    collection: Collection<RawDocumentBuf>,
+    cursor: Arc<Cursor<RawDocumentBuf>>,
```

Comment:
This is what's causing stuff to not be Sync.

What we can do here is wrap it in a mutex like so:

```
cursor: Mutex<Option<Cursor<RawDocumentBuf>>>
```

(This requires adding parking_lot to the cargo.toml)

Then in `execute`, we take the cursor out of the mutex:

```rust
    fn execute(
        &self,
        partition: usize,
        _context: Arc<TaskContext>,
    ) -> DatafusionResult<SendableRecordBatchStream> {
        if partition != 0 {
            return Err(DataFusionError::Execution(format!(
                "only single partition supported"
            )));
        }

        let cursor = {
            let mut cursor = self.cursor.lock();
            cursor.take().ok_or_else(|| {
                DataFusionError::Execution(format!(
                    "execution called on partition {partition} more than once"
                ))
            })?
        };

        Ok(Box::pin(DataSourceMetricsStreamAdapter::new(
            BsonStream::new(cursor, self.schema.clone(), self.limit),
            partition,
            &self.metrics,
        )))
    }
```

`execute` is only called once per partition at the very beginning of execution, and so the mutex is only getting locked once with no contention. The actual execution stream (what we return) remains lock free.

Also the `Arc` isn't needed here.

Reply:
It looks like the Arc is needed, because the cursor doesn't implement clone 😞

Reply:
We can remove the clone, it shouldn't be needed.

Reply:
I thought clone was passing when passing "by value"

Reply:
Clone just lets you make a copy of it via `clone()`. Everything is pass by value (objects and references).